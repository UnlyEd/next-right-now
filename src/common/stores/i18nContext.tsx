import React from 'react';

export type I18nContext = {
  locale: string; // e.g: fr, fr-FR, en, en-US, en-GB
  lang: string; // e.g: fr, en
}

/**
 * Uses native React Context API
 *
 * @example Usage
 *  import i18nContext from './src/stores/i18nContext';
 *  const { locale, lang }: I18nContext = React.useContext(i18nContext);
 *
 * @see https://reactjs.org/docs/context.html
 * @see https://medium.com/better-programming/react-hooks-usecontext-30eb560999f for useContext hook example (open in anonymous browser #paywall)
 */
export const i18nContext = React.createContext<I18nContext>(null);

export default i18nContext;
