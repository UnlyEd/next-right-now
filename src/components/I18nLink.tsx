import Link from 'next/link';
import React from 'react';
import { I18nRoute, resolveI18nRoute } from '../utils/router';

const I18nLink = (props: Props): JSX.Element => {
  const {
    as,
    children,
    href,
    locale,
    ...rest // Should only contain valid next/Link props
  } = props;
  const {
    i18nHref,
    i18nAs,
  }: I18nRoute = resolveI18nRoute({ as, href, locale });

  return (
    <Link
      href={i18nHref}
      as={i18nAs}
      {...rest}
    >
      {children}
    </Link>
  );
};

type Props = {
  children: React.ReactElement;
  locale: string;
  href: string;
  as?: string;
  replace?: boolean;
  scroll?: boolean;
  shallow?: boolean;
  passHref?: boolean;
  prefetch?: boolean;
}

export default I18nLink;
