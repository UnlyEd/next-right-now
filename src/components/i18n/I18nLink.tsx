import isArray from 'lodash.isarray';
import isEmpty from 'lodash.isempty';
import map from 'lodash.map';
import NextLink from 'next/link';
import React from 'react';
import useI18n, { I18n } from '../../hooks/useI18n';
import { I18nRoute, resolveI18nRoute } from '../../utils/app/router';

type ParamValueToForward = string | number | Array<string | number>;

type Props = {
  as?: string;
  children: React.ReactNode;
  className?: string;
  href: string;
  locale?: string; // The locale can be specified, but it'll fallback to the current locale if unspecified
  params?: { [key: string]: ParamValueToForward };
  passHref?: boolean;
  prefetch?: boolean;
  query?: { [key: string]: ParamValueToForward };
  replace?: boolean;
  scroll?: boolean;
  shallow?: boolean;
  wrapChildrenAsLink?: boolean; // Helper to avoid writing redundant code
};

/**
 * Wrapper around the native Next.js <Link> component. Handles localised links.
 *
 * Use the current active locale by default
 *
 * @example Recommended usage
 *  <I18nLink href={`/`}>Homepage</I18nLink>
 *  <I18nLink href={`/`} className="customClassName">Homepage</I18nLink>
 *
 * @example When specifying the locale to use (overrides default)
 *  <I18nLink href={`/`} locale={'fr-FR'}>Homepage in "fr-FR" locale</I18nLink>
 *
 * @example The recommended usage is equivalent to this (wrapChildrenAsLink is true, by default)
 *  <I18nLink href={`/`} wrapChildrenAsLink={false}><a>Homepage</a></I18nLink>
 *  <I18nLink href={`/`} wrapChildrenAsLink={false}><a className="customClassName">Homepage</a></I18nLink>
 *
 * @example When using children that use a <a> tag, you must set wrapChildrenAsLink to false to avoid using a link within a link
 *   <I18nLink
 *     href={`/`}
 *     wrapChildrenAsLink={false}
 *   >
 *      <NavLink>Homepage</NavLink>
 *   </I18nLink>
 *
 * @example When using route params (other than "locale"). Ex: /products/5
 *   <I18nLink
 *     href={'/products/[id]'}
 *     params={{
 *       id: 5,
 *     }}
 *   >
 *      Go to product 5
 *   </I18nLink>
 *
 * @example When using route query params. Ex: /products/5?userId=1
 *   <I18nLink
 *     href={'/products/[id]'}
 *     params={{
 *       id: 5,
 *     }}
 *     query={{
 *       userId: 1
 *     }}
 *   >
 *      Go to product 5 with userId 1
 *   </I18nLink>
 *
 * @param props
 */
const I18nLink: React.FunctionComponent<Props> = (props): JSX.Element => {
  const { locale: currentLocale }: I18n = useI18n();
  const {
    as,
    children,
    className,
    href,
    locale = currentLocale,
    params,
    passHref = true,
    query,
    wrapChildrenAsLink = true,
    ...rest // Should only contain valid next/Link props
  } = props;
  let {
    i18nHref, // eslint-disable-line prefer-const
    i18nAs,
  }: I18nRoute = resolveI18nRoute({ as, href, locale });

  if (!isEmpty(params)) {
    // If any params are provided, replace their name by the provided value
    map(params, (value: ParamValueToForward, key: string | number) => {
      if (isArray(value)) {
        value = value.join(',');
      }
      i18nAs = i18nAs.replace(`[${key}]`, value as string);
    });
  }

  if (!isEmpty(query)) {
    // If any query params are provided, append it to `as`, so it gets forwarded;
    const queryParamsString = Object.keys(query)
      .map((k) => {
        if (isArray(k)) {
          k = k.join(',');
        }
        return `${k}=${query[k]}`;
      })
      .join('&');
    i18nHref += `?${queryParamsString}`;
    i18nAs += `?${queryParamsString}`;
  }

  return (
    <NextLink
      href={i18nHref}
      as={i18nAs}
      {...rest}
      passHref={passHref}
    >
      {wrapChildrenAsLink ? (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a className={className}>{children}</a>
      ) : (
        React.cloneElement(children as React.ReactElement)
      )}
    </NextLink>
  );
};

export default I18nLink;
