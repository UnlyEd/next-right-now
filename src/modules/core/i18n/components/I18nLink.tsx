import isArray from 'lodash.isarray';
import isEmpty from 'lodash.isempty';
import map from 'lodash.map';
import NextLink from 'next/link';
import React from 'react';
import useI18n, { I18n } from '../hooks/useI18n';
import {
  I18nRoute,
  resolveI18nRoute,
} from '../i18nRouter';

type ParamValueToForward = string | number | Array<string | number>;

type Props = {
  /**
   * Optional decorator for the path that will be shown in the browser URL bar.
   */
  as?: string;

  /**
   * React node as children.
   */
  children: React.ReactNode;

  /**
   * Additional CSS classes.
   */
  className?: string;

  /**
   * The path or URL to navigate to.
   */
  href: string;

  /**
   * The active locale is automatically prepended. locale allows for providing a different locale.
   *
   * @default current locale
   */
  locale?: string; // The locale can be specified, but it'll fallback to the current locale if unspecified

  /**
   * Parameters to inject into the url, necessary when using route params (other than `locale`).
   *
   * Example:
   *
   * `/products/[id]` with `params={{ id: 5 }}` becomes `/products/5`
   */
  params?: { [key: string]: ParamValueToForward };

  /**
   * Forces Link to send the href property to its child.
   *
   * @default false
   */
  passHref?: boolean;

  /**
   * Prefetch the page in the background.
   * Any <Link /> that is in the viewport (initially or through scroll) will be preloaded.
   * Prefetch can be disabled by passing prefetch={false}.
   * Pages using Static Generation will preload JSON files with the data for faster page transitions.
   *
   * @default true
   */
  prefetch?: boolean;

  /**
   * Query to inject to the url, necessary when using route query param.
   *
   * Example:
   *
   * `/products` with `query={{ userId: 1 }}` becomes `/products?userId=1`
   */
  query?: { [key: string]: ParamValueToForward };

  /**
   * Replace the current history state instead of adding a new url into the stack.
   *
   * @default false
   */
  replace?: boolean;

  /**
   * Scroll to the top of the page after a navigation.
   *
   * @default true
   */
  scroll?: boolean;

  /**
   * Update the path of the current page without rerunning `getStaticProps`, `getServerSideProps` or `getInitialProps`.
   *
   * @default false
   */
  shallow?: boolean;

  /**
   * Disabled on Storybook, as it crashes the UI.
   */
  wrapChildrenAsLink?: boolean; // Helper to avoid writing redundant code
};

/**
 * Wrapper around the native Next.js <Link> component. Handles localised links.
 *
 * Uses the current active locale by default.
 *
 * [Read why we don't use the official Next.js `Link` component](https://unlyed.github.io/next-right-now/guides/i18n/#official-i18n-routing-implementation) <span className="tip ml-2">Tip</span>
 *
 * @example Recommended usage
 *  <I18nLink href={`/`}>Homepage</I18nLink>
 *  <I18nLink href={`/`} className="customClassName">Homepage</I18nLink>
 *
 * @example When specifying the locale to use (overrides default)
 *  <I18nLink href={`/`} locale={'fr-FR'}>Homepage in "fr-FR" locale</I18nLink>
 *
 * @example The recommended usage is equivalent to this (wrapChildrenAsLink is true, by default)
 *  <I18nLink href={`/`} wrapChildrenAsLink={false}><a>Homepage</a></I18nLink>
 *  <I18nLink href={`/`} wrapChildrenAsLink={false}><a className="customClassName">Homepage</a></I18nLink>
 *
 * @example When using children that use a <a> tag, you must set wrapChildrenAsLink to false to avoid using a link within a link
 *   <I18nLink
 *     href={`/`}
 *     wrapChildrenAsLink={false}
 *   >
 *      <NavLink>Homepage</NavLink>
 *   </I18nLink>
 *
 * @example When using route params (other than "locale"). Ex: /products/5
 *   <I18nLink
 *     href={'/products/[id]'}
 *     params={{
 *       id: 5,
 *     }}
 *   >
 *      Go to product 5
 *   </I18nLink>
 *
 * @example When using route query params. Ex: /products/5?userId=1
 *   <I18nLink
 *     href={'/products/[id]'}
 *     params={{
 *       id: 5,
 *     }}
 *     query={{
 *       userId: 1
 *     }}
 *   >
 *      Go to product 5 with userId 1
 *   </I18nLink>
 *
 * @param props
 */
const I18nLink: React.FunctionComponent<Props> = (props): JSX.Element => {
  const { locale: currentLocale }: I18n = useI18n();
  const {
    as,
    children,
    className,
    href,
    locale = currentLocale,
    params,
    passHref = true,
    query,
    wrapChildrenAsLink = true,
    ...rest // Should only contain valid next/Link props
  } = props;
  let {
    i18nHref, // eslint-disable-line prefer-const
    i18nAs,
  }: I18nRoute = resolveI18nRoute({
    as,
    href,
    locale,
  });

  if (!isEmpty(params)) {
    // If any params are provided, replace their name by the provided value
    map(params, (value: ParamValueToForward, key: string | number) => {
      if (isArray(value)) {
        value = value.join(',');
      }
      i18nAs = i18nAs.replace(`[${key}]`, value as string);
    });
  }

  if (!isEmpty(query)) {
    // If any query params are provided, append it to `as`, so it gets forwarded;
    const queryParamsString = Object.keys(query)
      .map((k) => {
        if (isArray(k)) {
          k = k.join(',');
        }
        return `${k}=${query[k]}`;
      })
      .join('&');
    i18nHref += `?${queryParamsString}`;
    i18nAs += `?${queryParamsString}`;
  }

  return (
    <NextLink
      href={i18nHref}
      as={i18nAs}
      {...rest}
      passHref={passHref}
    >
      {wrapChildrenAsLink ? (
        // eslint-disable-next-line jsx-a11y/anchor-is-valid
        <a className={className}>{children}</a>
      ) : (
        React.cloneElement(children as React.ReactElement)
      )}
    </NextLink>
  );
};

export default I18nLink;
