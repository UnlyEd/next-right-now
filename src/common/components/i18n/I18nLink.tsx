import Link from 'next/link';
import React from 'react';
import i18nContext, { I18nContext } from '../../stores/i18nContext';
import { I18nRoute, resolveI18nRoute } from '../../utils/app/router';

const I18nLink: React.FunctionComponent<Props> = (props): JSX.Element => {
  const {
    as,
    children,
    href,
    ...rest // Should only contain valid next/Link props
  } = props;
  const { locale }: I18nContext = React.useContext(i18nContext);
  const {
    i18nHref,
    i18nAs,
  }: I18nRoute = resolveI18nRoute({ as, href, locale });

  return (
    <Link
      href={i18nHref}
      as={i18nAs}
      {...rest}
    >
      {children}
    </Link>
  );
};

type Props = {
  children: React.ReactElement;
  href: string;
  as?: string;
  replace?: boolean;
  scroll?: boolean;
  shallow?: boolean;
  passHref?: boolean;
  prefetch?: boolean;
}

export default I18nLink;
