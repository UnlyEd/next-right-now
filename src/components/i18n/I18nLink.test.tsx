import React from 'react';
import TestRenderer from 'react-test-renderer';
import i18nContext from '../../stores/i18nContext';
import I18nLink from './I18nLink';
import { NavLink } from 'reactstrap';

/**
 * @group unit
 * @group components
 */
describe('I18nLink', () => {
  const I18nLinkTest = (props) => {
    const { locale = 'en', href, text = 'Text', ...rest } = props;

    return (
      <i18nContext.Provider value={{ lang: null, locale: locale }}>
        <I18nLink
          href={href}
          {...rest}
        >
          {text}
        </I18nLink>
      </i18nContext.Provider>
    );
  };

  describe('should properly render', () => {
    test('when going to homepage, it should redirect to i18n homepage', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/'} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/en');
      expect(link).toMatchSnapshot();
    });

    test('when using custom CSS class', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/'} className="customClassName" />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/en');
      expect(link.props.className).toEqual('customClassName');
      expect(link).toMatchSnapshot();
    });

    test('when forcing the locale to use', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/'} locale={'fr-FR'} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/fr-FR');
      expect(link).toMatchSnapshot();
    });

    test('when using wrapChildrenAsLink manually using a <a> element', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/'} wrapChildrenAsLink={false} text={<a>Page</a>} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Page']);
      expect(link.props.href).toEqual('/en');
      expect(link).toMatchSnapshot();
    });

    test('when using wrapChildrenAsLink manually using a <NavLink> element', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/'} wrapChildrenAsLink={false} text={<NavLink>Page</NavLink>} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Page']);
      expect(link.props.href).toEqual('/en');
      expect(link.props.className).toEqual('nav-link');
      expect(link).toMatchSnapshot();
    });

    test('when using route params', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/products/[id]'} params={{id: 5}} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/en/products/5');
      expect(link).toMatchSnapshot();
    });

    test('when using route params and query route params', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/products/[id]'} params={{id: 5}} query={{userId: 1}} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/en/products/5?userId=1');
      expect(link).toMatchSnapshot();
    });

    test('when using route params and query route params using nested paths', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/products/favourites/[id]'} params={{id: 5}} query={{userId: 1}} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/en/products/favourites/5?userId=1');
      expect(link).toMatchSnapshot();
    });

    test('when using route params and query route params using nested paths and forcing locale', () => {
      const renderer = TestRenderer.create(<I18nLinkTest href={'/products/favourites/[id]'} params={{id: 5}} query={{userId: 1}} locale={'fr'} />);
      const link: any = renderer.toJSON();

      expect(link.type).toEqual('a');
      expect(link.children).toEqual(['Text']);
      expect(link.props.href).toEqual('/fr/products/favourites/5?userId=1');
      expect(link).toMatchSnapshot();
    });
  });
});
