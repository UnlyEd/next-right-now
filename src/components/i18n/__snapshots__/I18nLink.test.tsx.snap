// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18nLink should properly render when forcing the locale to use 1`] = `
<a
  href="/fr-FR"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when going to homepage, it should redirect to i18n homepage 1`] = `
<a
  href="/en"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when using custom CSS class 1`] = `
<a
  className="customClassName"
  href="/en"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when using route params 1`] = `
<a
  href="/en/products/5"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when using route params and query route params 1`] = `
<a
  href="/en/products/5?userId=1"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when using route params and query route params using nested paths 1`] = `
<a
  href="/en/products/favourites/5?userId=1"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when using route params and query route params using nested paths and forcing locale 1`] = `
<a
  href="/fr/products/favourites/5?userId=1"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Text
</a>
`;

exports[`I18nLink should properly render when using wrapChildrenAsLink manually using a <NavLink> element 1`] = `
<a
  className="nav-link"
  href="/en"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Page
</a>
`;

exports[`I18nLink should properly render when using wrapChildrenAsLink manually using a <a> element 1`] = `
<a
  href="/en"
  onClick={[Function]}
  onMouseEnter={[Function]}
>
  Page
</a>
`;
